internal/resolvers/wikipedia/helpers.go

```diff
@@ -0,0 +1,93 @@
+package wikipedia
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Chatterino/api/pkg/cache"
+	"github.com/Chatterino/api/pkg/resolver"
+	"github.com/Chatterino/api/pkg/utils"
+)
+
+func getPageInfo(urlString string) (*wikipediaTooltipData, error) {
+	u, err := url.Parse(urlString)
+	if err != nil {
+		return nil, err
+	}
+
+	// Since the Wikipedia API is locale-dependant, we need the locale code.
+	// For example, if you want to resolve a de.wikipedia.org link, you need
+	// to ping the DE API endpoint.
+	localeMatch := localeRegexp.FindStringSubmatch(u.Hostname())
+	if len(localeMatch) != 2 {
+		return nil, errLocaleMatch
+	}
+
+	localeCode := localeMatch[1]
+
+	titleMatch := titleRegexp.FindStringSubmatch(u.Path)
+	if len(titleMatch) != 2 {
+		return nil, errTitleMatch
+	}
+
+	canonicalName := titleMatch[1]
+
+	requestURL := fmt.Sprintf(endpointURL, localeCode, canonicalName)
+
+	resp, err := resolver.RequestGET(requestURL)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("bad status: %d", resp.StatusCode)
+	}
+
+	var pageInfo *wikipediaAPIResponse
+	if err = json.NewDecoder(resp.Body).Decode(&pageInfo); err != nil {
+		return nil, err
+	}
+
+	// Transform API response into our tooltip model for Wikipedia links
+	tooltipData := &wikipediaTooltipData{}
+
+	tooltipData.Title = utils.TruncateString(pageInfo.Titles.Display, maxTitleLength)
+	tooltipData.Extract = utils.TruncateString(pageInfo.Extract, maxExtractLength)
+
+	if pageInfo.Description != nil {
+		tooltipData.Description = utils.TruncateString(*pageInfo.Description, maxDescriptionLength)
+	} else {
+		tooltipData.Description = ""
+	}
+
+	if pageInfo.Thumbnail != nil {
+		tooltipData.ThumbnailURL = pageInfo.Thumbnail.URL
+	}
+
+	return tooltipData, nil
+}
+
+func buildTooltip(pageInfo *wikipediaTooltipData) (interface{}, time.Duration, error) {
+	var tooltip bytes.Buffer
+
+	if err := wikipediaTooltipTemplate.Execute(&tooltip, pageInfo); err != nil {
+		return &resolver.Response{
+			Status:  http.StatusInternalServerError,
+			Message: "Wikipedia template error: " + resolver.CleanResponse(err.Error()),
+		}, cache.NoSpecialDur, nil
```

Comment:
we should return a `response` here

see if `buildTooltip` can be changed to return `(response, time.Duration, error)` instead (not sure if go would allow this

Reply:
Addressed in 5d5e233